From 8f70485f3d9c6f5ee4beb1b57d2cbf9f68b06b0d Mon Sep 17 00:00:00 2001
From: Hyun Kwon <hyun.kwon@xilinx.com>
Date: Thu, 4 Apr 2019 10:03:32 -0700
Subject: [PATCH 0615/1852] uio: xilinx_ai_engine: Enable the irq

commit dd0cc93c01f976b9dff478e4a5fc63cdd71849d8 from
https://github.com/Xilinx/linux-xlnx.git

Pass the irq info from dt to the uio genirq. For now, only enable
one interrupt, but there's can be up to 3 interrupts assigned.

Signed-off-by: Hyun Kwon <hyun.kwon@xilinx.com>
Reviewed-by: Wendy Liang <wendy.liang@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/uio/uio_xilinx_ai_engine.c | 21 ++++++++++++++++++++-
 1 file changed, 20 insertions(+), 1 deletion(-)

diff --git a/drivers/uio/uio_xilinx_ai_engine.c b/drivers/uio/uio_xilinx_ai_engine.c
index 96d0e67f23a6..29538f770d1f 100644
--- a/drivers/uio/uio_xilinx_ai_engine.c
+++ b/drivers/uio/uio_xilinx_ai_engine.c
@@ -16,6 +16,7 @@
 #include <linux/uio_driver.h>
 
 #define DRIVER_NAME "xilinx-aiengine"
+#define XILINX_AI_ENGINE_MAX_IRQ	4
 
 static uint xilinx_ai_engine_mem_cnt = 1;
 module_param_named(mem_cnt, xilinx_ai_engine_mem_cnt, uint, 0444);
@@ -89,6 +90,10 @@ static int xilinx_ai_engine_probe(struct platform_device *pdev)
 	struct platform_device *uio;
 	struct uio_dmem_genirq_pdata *pdata;
 	unsigned int i;
+	static const char * const interrupt_names[] = { "interrupt0",
+							"interrupt1",
+							"interrupt2",
+							"interrupt3" };
 	int ret;
 
 	uio = platform_device_alloc(DRIVER_NAME, PLATFORM_DEVID_NONE);
@@ -107,11 +112,25 @@ static int xilinx_ai_engine_probe(struct platform_device *pdev)
 	pdata->dynamic_region_sizes = &xilinx_ai_engine_mem_size;
 	pdata->uioinfo.name = DRIVER_NAME;
 	pdata->uioinfo.version = "devicetree";
-	pdata->uioinfo.irq = UIO_IRQ_CUSTOM;
 	pdata->uioinfo.mmap = xilinx_ai_engine_mmap;
 	/* Set the offset value as it's map index for each memory */
 	for (i = 0; i < MAX_UIO_MAPS; i++)
 		pdata->uioinfo.mem[i].offs = i << PAGE_SHIFT;
+
+	/* TODO: Only one interrupt is supported out of 4 */
+	for (i  = 0; i < XILINX_AI_ENGINE_MAX_IRQ; i++) {
+		ret = platform_get_irq_byname(pdev, interrupt_names[i]);
+		if (ret >= 0) {
+			dev_info(&pdev->dev, "%s is used", interrupt_names[i]);
+			break;
+		}
+	}
+
+	/* Interrupt is optional */
+	if (ret < 0)
+		ret = UIO_IRQ_CUSTOM;
+	pdata->uioinfo.irq = ret;
+
 	ret = platform_device_add_data(uio, pdata, sizeof(*pdata));
 	if (ret)
 		goto err_out;
-- 
2.31.1

